server/src/main/java/org/elasticsearch/gateway/WriteStateException.java

```diff
@@ -38,4 +40,15 @@
     public boolean isDirty() {
         return dirty;
     }
+
+    public void rethrowAsErrorOrUncheckedException(String msg) {
+        if (isDirty()) {
+            // IOError is the best thing we have in java library to indicate that serious IO problem has occurred.
+            // IOError will be caught by ElasticsearchUncaughtExceptionHandler and JVM will be halted.
+            // Sadly, it has no constructor that accepts error message, so we first wrap WriteStateException with IOException.
+            throw new IOError(new IOException(msg + ", storage is dirty", this));
```

Comment:
I think we do not need the message here: the message indicates the calling method but this is also in the stack trace (and the previous log line) and we would know that the storage was dirty if we have an `IOError` caused by a `WriteStateException`.